File: packages/attachment_engine/attachment_engine_macos/macos/attachment_engine_macos/Sources/attachment_engine_macos/VideoPlatformView.swift
import AVFoundation
import AVKit
import Cocoa
import FlutterMacOS

/// Embeds an `AVPlayer`-backed `AVPlayerLayer` through a
/// `FlutterPlatformViewFactory` (macOS's platform-view embedding API is
/// simpler than iOS's — no `FlutterPlatformView` wrapper protocol; the
/// factory returns the `NSView` directly, and there is no `frame`
/// parameter to `createWithViewIdentifier`), mirroring the Android
/// `TextureView`/`MediaPlayer` implementation: control
/// (`load`/`play`/`pause`/`seek`/`setSpeed`/`setVolume`/`dispose`) goes
/// over `attachment_engine/video`, and playback/buffering events go over
/// `attachment_engine/video_events/{playerId}`.
class VideoPlayerEntry: NSObject {
  let playerId: String
  let player = AVPlayer()
  var eventSink: FlutterEventSink?
  private var eventChannel: FlutterEventChannel?
  private var progressTimer: Timer?
  private var itemObservation: NSKeyValueObservation?

  init(playerId: String, messenger: FlutterBinaryMessenger) {
    self.playerId = playerId
    super.init()
    eventChannel = FlutterEventChannel(
      name: "attachment_engine/video_events/\(playerId)", binaryMessenger: messenger)
    eventChannel?.setStreamHandler(
      VideoStreamHandler(onListen: { [weak self] sink in
        self?.eventSink = sink
      }, onCancel: { [weak self] in
        self?.eventSink = nil
      }))
  }

  func load(path: String?, url: String?) {
    stopTimer()
    emit(state: "buffering")
    let mediaUrl: URL?
    if let path = path {
      mediaUrl = URL(fileURLWithPath: path)
    } else if let url = url {
      mediaUrl = URL(string: url)
    } else {
      mediaUrl = nil
    }
    guard let mediaUrl = mediaUrl else {
      emit(state: "error")
      return
    }
    let item = AVPlayerItem(url: mediaUrl)
    NotificationCenter.default.addObserver(
      self, selector: #selector(didFinishPlaying), name: .AVPlayerItemDidPlayToEndTime,
      object: item)
    itemObservation = item.observe(\.status, options: [.new]) { [weak self] item, _ in
      guard let self = self else { return }
      switch item.status {
      case .readyToPlay:
        self.emit(state: "ready")
        self.startTimer()
      case .failed:
        self.emit(state: "error")
      default:
        break
      }
    }
    player.replaceCurrentItem(with: item)
  }

  @objc private func didFinishPlaying() {
    emit(state: "completed")
  }

  func play() {
    player.play()
    emit(state: "playing")
  }

  func pause() {
    player.pause()
    emit(state: "paused")
  }

  func seek(ms: Int) {
    player.seek(to: CMTime(seconds: Double(ms) / 1000.0, preferredTimescale: 1000))
  }

  func setSpeed(_ speed: Float) {
    player.rate = speed
  }

  func setVolume(_ volume: Float) {
    player.volume = volume
  }

  private func startTimer() {
    stopTimer()
    let timer = Timer.scheduledTimer(withTimeInterval: 0.25, repeats: true) { [weak self] _ in
      self?.emitProgress()
    }
    RunLoop.main.add(timer, forMode: .common)
    progressTimer = timer
  }

  private func stopTimer() {
    progressTimer?.invalidate()
    progressTimer = nil
  }

  private func emitProgress() {
    guard let item = player.currentItem else { return }
    let positionMs = Int(player.currentTime().seconds * 1000)
    let durationSeconds = item.duration.seconds
    let durationMs = durationSeconds.isFinite ? Int(durationSeconds * 1000) : nil
    var event: [String: Any] = [
      "state": player.rate > 0 ? "playing" : "paused",
      "positionMs": positionMs,
    ]
    if let durationMs = durationMs { event["durationMs"] = durationMs }
    if let track = item.asset.tracks(withMediaType: .video).first {
      let size = track.naturalSize.applying(track.preferredTransform)
      event["width"] = abs(size.width)
      event["height"] = abs(size.height)
    }
    eventSink?(event)
  }

  private func emit(state: String) {
    eventSink?(["state": state])
  }

  func dispose() {
    stopTimer()
    player.pause()
    itemObservation?.invalidate()
    NotificationCenter.default.removeObserver(self)
    eventChannel?.setStreamHandler(nil)
  }
}

private class VideoStreamHandler: NSObject, FlutterStreamHandler {
  let onListen: (FlutterEventSink?) -> Void
  let onCancel: () -> Void

  init(onListen: @escaping (FlutterEventSink?) -> Void, onCancel: @escaping () -> Void) {
    self.onListen = onListen
    self.onCancel = onCancel
  }

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    onListen(events)
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    onCancel()
    return nil
  }
}

/// Shared registry so the control API and the platform view factory both
/// operate on the same native player instance for a given `playerId`.
///
/// Method calls (load/play/pause/seek/setSpeed/setVolume/dispose) are
/// dispatched through the Pigeon-generated `VideoHostApi`. The platform
/// view factory below stays hand-registered (it isn't a method call), and
/// playback-state *events* stay on the hand-written per-`playerId`
/// `FlutterEventChannel` above (`VideoPlayerEntry`/`VideoStreamHandler`) —
/// see the note in `attachment_engine_platform_interface/pigeons/messages.dart`.
class VideoChannel: NSObject, VideoHostApi {
  static let viewType = "attachment_engine/video_view"

  private var messenger: FlutterBinaryMessenger!
  var entries: [String: VideoPlayerEntry] = [:]

  func register(with messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    VideoHostApiSetup.setUp(binaryMessenger: messenger, api: self)
  }

  func unregister(with messenger: FlutterBinaryMessenger) {
    VideoHostApiSetup.setUp(binaryMessenger: messenger, api: nil)
    for entry in entries.values { entry.dispose() }
    entries.removeAll()
  }

  func entry(for playerId: String) -> VideoPlayerEntry {
    if let existing = entries[playerId] { return existing }
    let entry = VideoPlayerEntry(playerId: playerId, messenger: messenger)
    entries[playerId] = entry
    return entry
  }

  @MainActor
  func load(playerId: String, filePath: String?, url: String?) async throws {
    entry(for: playerId).load(path: filePath, url: url)
  }

  @MainActor
  func play(playerId: String) async throws {
    entries[playerId]?.play()
  }

  @MainActor
  func pause(playerId: String) async throws {
    entries[playerId]?.pause()
  }

  @MainActor
  func seek(playerId: String, positionMs: Int64) async throws {
    entries[playerId]?.seek(ms: Int(positionMs))
  }

  @MainActor
  func setSpeed(playerId: String, speed: Double) async throws {
    entries[playerId]?.setSpeed(Float(speed))
  }

  @MainActor
  func setVolume(playerId: String, volume: Double) async throws {
    entries[playerId]?.setVolume(Float(volume))
  }

  @MainActor
  func dispose(playerId: String) async throws {
    entries.removeValue(forKey: playerId)?.dispose()
  }
}

/// macOS's `FlutterPlatformViewFactory` has no `FlutterPlatformView`
/// wrapper protocol and no `frame` parameter — it returns the `NSView`
/// directly.
class VideoPlatformViewFactory: NSObject, FlutterPlatformViewFactory {
  private let videoChannel: VideoChannel

  init(videoChannel: VideoChannel) {
    self.videoChannel = videoChannel
    super.init()
  }

  func create(withViewIdentifier viewId: Int64, arguments args: Any?) -> NSView {
    let params = args as? [String: Any]
    let playerId = (params?["playerId"] as? String) ?? String(viewId)
    return VideoPlayerNSView(entry: videoChannel.entry(for: playerId))
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    FlutterStandardMessageCodec.sharedInstance()
  }
}

private class VideoPlayerNSView: NSView {
  private let playerLayer: AVPlayerLayer

  init(entry: VideoPlayerEntry) {
    playerLayer = AVPlayerLayer(player: entry.player)
    playerLayer.videoGravity = .resizeAspect
    super.init(frame: .zero)
    wantsLayer = true
    layer?.backgroundColor = NSColor.black.cgColor
    layer?.addSublayer(playerLayer)
  }

  required init?(coder: NSCoder) {
    fatalError("init(coder:) is not supported")
  }

  override func layout() {
    super.layout()
    playerLayer.frame = bounds
  }
}

File: packages/attachment_engine/attachment_engine_android/android/src/main/kotlin/com/dhc/attachment_engine/VideoPlatformView.kt
// Copyright 2026 DHC Tech
// Use of this source code is governed by an MIT-style license that can be
// found in the LICENSE file.

package com.dhc.attachment_engine

import android.content.Context
import android.graphics.SurfaceTexture
import android.media.MediaPlayer
import android.os.Handler
import android.os.Looper
import android.view.Surface
import android.view.TextureView
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

/**
 * Replaces `video_player`. Embeds a [TextureView] backed by [android.media.MediaPlayer]
 * (Media3/ExoPlayer would be preferable for adaptive streaming, but was intentionally avoided here
 * to keep this pass free of additional third-party Gradle dependencies — see README "Known
 * limitations": no adaptive bitrate/DASH/HLS support).
 */
class VideoPlayerEntry(val playerId: String, private val messenger: BinaryMessenger) {
  val mediaPlayer = MediaPlayer()
  var surface: Surface? = null
  var texture: SurfaceTexture? = null
  private val handler = Handler(Looper.getMainLooper())
  private var progressRunnable: Runnable? = null
  var eventSink: EventChannel.EventSink? = null
  private val eventChannel =
      EventChannel(messenger, "attachment_engine/video_events/$playerId").also {
        it.setStreamHandler(
            object : EventChannel.StreamHandler {
              override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
              }

              override fun onCancel(arguments: Any?) {
                eventSink = null
              }
            },
        )
      }

  fun attachSurface(newSurface: Surface) {
    surface = newSurface
    mediaPlayer.setSurface(newSurface)
  }

  fun load(path: String?, url: String?) {
    try {
      mediaPlayer.reset()
      mediaPlayer.setDataSource(path ?: url ?: "")
      surface?.let { mediaPlayer.setSurface(it) }
      mediaPlayer.setOnPreparedListener {
        emit("ready")
        startProgressLoop()
      }
      mediaPlayer.setOnCompletionListener { emit("completed") }
      mediaPlayer.setOnErrorListener { _, _, _ ->
        emit("error")
        true
      }
      emit("buffering")
      mediaPlayer.prepareAsync()
    } catch (_: Exception) {
      emit("error")
    }
  }

  fun play() {
    mediaPlayer.start()
    emit("playing")
  }

  fun pause() {
    mediaPlayer.pause()
    emit("paused")
  }

  fun seek(ms: Int) = mediaPlayer.seekTo(ms)

  fun setSpeed(speed: Float) {
    try {
      mediaPlayer.playbackParams = mediaPlayer.playbackParams.setSpeed(speed)
    } catch (_: Exception) {}
  }

  fun setVolume(volume: Float) = mediaPlayer.setVolume(volume, volume)

  private fun startProgressLoop() {
    progressRunnable?.let { handler.removeCallbacks(it) }
    val runnable =
        object : Runnable {
          override fun run() {
            try {
              eventSink?.success(
                  mapOf(
                      "state" to if (mediaPlayer.isPlaying) "playing" else "paused",
                      "positionMs" to mediaPlayer.currentPosition,
                      "durationMs" to mediaPlayer.duration,
                      "width" to mediaPlayer.videoWidth,
                      "height" to mediaPlayer.videoHeight,
                  ),
              )
            } catch (_: Exception) {}
            handler.postDelayed(this, 250)
          }
        }
    progressRunnable = runnable
    handler.post(runnable)
  }

  private fun emit(state: String) {
    eventSink?.success(mapOf("state" to state))
  }

  fun dispose() {
    progressRunnable?.let { handler.removeCallbacks(it) }
    eventChannel.setStreamHandler(null)
    try {
      mediaPlayer.release()
    } catch (_: Exception) {}
    texture?.release()
  }
}

/**
 * Shared registry so the control [MethodChannel] and the [PlatformView] factory (created lazily by
 * the Flutter engine when the widget attaches) both operate on the same native player instance for
 * a given `playerId`.
 */
/**
 * Method calls (load/play/pause/seek/setSpeed/setVolume/dispose) are dispatched through the
 * Pigeon-generated [VideoHostApi]. The platform-view factory below stays hand-registered (it isn't
 * a method call), and playback-state *events* stay on a hand-written per-`playerId` `EventChannel`
 * for the same reason as [AudioChannel] — see the note in
 * `attachment_engine_platform_interface/pigeons/messages.dart`.
 */
class VideoChannel(private val messenger: BinaryMessenger) : VideoHostApi {
  companion object {
    const val VIEW_TYPE = "attachment_engine/video_view"
  }

  val entries = mutableMapOf<String, VideoPlayerEntry>()

  fun register() {
    VideoHostApi.setUp(messenger, this)
  }

  fun unregister() {
    VideoHostApi.setUp(messenger, null)
    entries.keys.toList().forEach { disposeEntry(it) }
  }

  fun entryFor(playerId: String): VideoPlayerEntry {
    return entries.getOrPut(playerId) { VideoPlayerEntry(playerId, messenger) }
  }

  private fun disposeEntry(playerId: String) {
    entries.remove(playerId)?.dispose()
  }

  override suspend fun load(playerId: String, filePath: String?, url: String?) {
    entryFor(playerId).load(filePath, url)
  }

  override suspend fun play(playerId: String) {
    entries[playerId]?.play()
  }

  override suspend fun pause(playerId: String) {
    entries[playerId]?.pause()
  }

  override suspend fun seek(playerId: String, positionMs: Long) {
    entries[playerId]?.seek(positionMs.toInt())
  }

  override suspend fun setSpeed(playerId: String, speed: Double) {
    entries[playerId]?.setSpeed(speed.toFloat())
  }

  override suspend fun setVolume(playerId: String, volume: Double) {
    entries[playerId]?.setVolume(volume.toFloat())
  }

  override suspend fun dispose(playerId: String) {
    disposeEntry(playerId)
  }
}

class VideoPlatformViewFactory(private val videoChannel: VideoChannel) :
    PlatformViewFactory(StandardMessageCodec.INSTANCE) {
  override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
    @Suppress("UNCHECKED_CAST") val params = args as? Map<String, Any?>
    val playerId = params?.get("playerId") as? String ?: viewId.toString()
    return VideoPlatformViewImpl(context, videoChannel.entryFor(playerId))
  }
}

private class VideoPlatformViewImpl(context: Context, private val entry: VideoPlayerEntry) :
    PlatformView, TextureView.SurfaceTextureListener {
  private val textureView = TextureView(context)

  init {
    textureView.surfaceTextureListener = this
  }

  override fun getView(): View = textureView

  override fun onSurfaceTextureAvailable(surface: SurfaceTexture, width: Int, height: Int) {
    entry.texture = surface
    entry.attachSurface(Surface(surface))
  }

  override fun onSurfaceTextureSizeChanged(surface: SurfaceTexture, width: Int, height: Int) = Unit

  override fun onSurfaceTextureDestroyed(surface: SurfaceTexture): Boolean = true

  override fun onSurfaceTextureUpdated(surface: SurfaceTexture) = Unit

  override fun dispose() {
    // The native player entry outlives the view (pooled by Dart-side
    // VideoControllerPool); only the view's own resources are torn
    // down here.
  }
}
